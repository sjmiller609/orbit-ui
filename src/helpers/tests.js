import { BrowserRouter } from 'react-router-dom'
import { shape } from 'prop-types'
import { mount } from 'enzyme'

// Instantiate router context
const router = {
  history: new BrowserRouter().history,
  route: {
    location: {},
    match: {},
  },
}

const createContext = () => ({
  context: { router },
  childContextTypes: { router: shape({}) },
})

// wrap with router context - to mount RR links correctly
export function mountWrap(node) {
  return mount(node, createContext())
}
