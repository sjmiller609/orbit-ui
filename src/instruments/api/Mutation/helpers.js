const has = (str, find) => ~str.toLowerCase().indexOf(find.toLowerCase())

export const formErrors = ({ error, errors }) => {
  if (!error) return

  let err = JSON.stringify(error).toLowerCase()
  return errors.find(e => {
    if (e.key && has(err, e.key)) return e
    else if (e.allKeys && e.allKeys.every(k => has(err, k))) return e
    else if (e.someKeys && e.someKeys.some(k => has(err, k))) return e
  })
}
