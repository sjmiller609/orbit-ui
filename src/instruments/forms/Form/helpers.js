// faltten and stringify nested keys
export const unpack = (obj, ns, hash) => {
  if (!obj) return
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keypath = ns ? `${ns}.${key}` : key
    if (value && typeof value === 'object') {
      return unpack(value, keypath, acc)
    }
    acc[keypath] = value
    return acc
  }, hash || {})
}

const roll = (keys, obj, value) => {
  const key = keys[0]
  // case 1 - not obj
  if (keys.length === 1) {
    obj[key] = value
    return obj
  }
  // case 2 - doesn't exist
  if (!obj[key]) {
    obj[key] = isNaN(keys[1]) ? {} : []
  }
  // case 3 - exists
  obj[key] = roll(
    keys.slice(1),
    Array.isArray(obj[key]) ? [...obj[key]] : { ...obj[key] },
    value
  )

  return obj
}

// convert back into obj
export const pack = (obj, type = {}) => {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keys = key.split('.')

    return roll(keys, acc, value)
  }, type)
}

export const packChild = ({ name, obj }) => {
  const obj2 = {}
  let isArray
  Object.entries(obj).forEach(k => {
    if (name.length < k[0].length && k[0].indexOf(name) === 0) {
      const n = k[0].slice(name.length + 1)
      // must be true for all matching
      isArray = !isNaN(n.slice(0, n.indexOf('.')))
      obj2[n] = k[1]
    }
  })
  if (Object.keys(obj2).length) return pack(obj2, isArray ? [] : {})
  return null
}

export const removeChild = ({ name, obj = {} }) => {
  const obj2 = {}
  Object.entries(obj).forEach(k => {
    if (name.length >= k[0].length || k[0].indexOf(name) !== 0) {
      obj2[k[0]] = k[1]
    }
  })
  return obj2
}
