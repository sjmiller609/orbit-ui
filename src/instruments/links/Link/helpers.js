// check if url is external
export const externalUrl = url => {
  if (typeof url === 'object') return false
  const path = /^https?:\/\/|^\/\//i
  if (path.test(url) && !~url.indexOf(window.location.origin)) return true
  return
}
