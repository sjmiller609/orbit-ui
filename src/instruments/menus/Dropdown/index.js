'use strict'
import React from 'react'
import PropTypes from 'prop-types'
import s from './styles.scss'
import { Backdrop, Link, Card } from '../../../instruments'

import classnames from 'classnames'

class Dropdown extends React.Component {
  timeout = null
  blur = this.blur.bind(this)
  open = this.open.bind(this)

  state = {
    menu: false,
    out: true,
  }

  componentWillUpdate(nextProps, nextState) {
    if (this.state.menu && !nextState.menu)
      this.timeout = setTimeout(() => this.setState({ out: true }), 250)
  }
  componentWillUnmount() {
    clearTimeout(this.timeout)
  }

  blur() {
    this.setState({ menu: false })
  }

  open() {
    const { menu } = this.state
    if (menu) return
    this.setState({ menu: true, out: false })
  }

  render() {
    const { selector, disable, right, children, className } = this.props
    const { menu, out } = this.state

    return (
      <div
        className={classnames(
          s.dropdown,
          right ? s.right : s.left,
          menu ? s.open : null,
          out ? s.out : null,
          className
        )}>
        <Backdrop show={menu} close={this.blur} />
        {!disable ? (
          <Link onClick={!menu ? this.open : this.blur} className={s.selector}>
            {selector}
          </Link>
        ) : (
          selector
        )}
        <div className={s.menuWrapper}>
          <Card className={s.menu}>
            {Array.isArray(children) ? children.map(el => el) : children}
          </Card>
        </div>
      </div>
    )
  }
}

Dropdown.propTypes = {
  right: PropTypes.bool,
  disable: PropTypes.bool,
  selector: PropTypes.element.isRequired,
  className: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.array, PropTypes.element])
    .isRequired,
}

export default Dropdown
