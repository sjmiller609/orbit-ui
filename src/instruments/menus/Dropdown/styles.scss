@import 'src/styles/mixins.scss';

$width: 19rem;

.dropdown {
  position: relative;
}

.open {
  z-index: 1000;
}

.open .menuWrapper {
  opacity: 1;
  pointer-events: auto;
  z-index: 1000;
}

.menuWrapper {
  opacity: 0;
  position: absolute;
  width: $width + 2rem; //fix
  height: auto;
  margin-top: 0.7rem;
  transition: opacity 200ms ease-in-out;

  &:before {
    content: '';
    display: block;
    background: $white;
    height: 1rem;
    width: 1rem;
    transform: translateX(-50%) rotate(45deg);
    position: absolute;
    border-radius: 2px;
    border-left: 1px solid $shadow;
    border-top: 1px solid $shadow;
    top: -0.5rem;
    z-index: 2000;
  }
}

div.menu {
  position: relative;
  width: 100%;
  max-width: 48rem;
  height: auto;
  min-height: auto;
  max-height: 70vh;
  overflow-y: auto;
  margin: 0;
}

.out .menuWrapper {
  z-index: -1;
  height: 0;
}

.selector {
  z-index: inherit;
  position: relative;
}

.right .menuWrapper {
  right: -0.5rem;
  &:before {
    right: 1.1rem;
  }
}
.left .menuWrapper {
  left: -0.5rem;
  &:before {
    left: 3rem;
  }
}

.small .menuWrapper {
  width: auto;
  margin-top: 0.5rem;

  &:before {
    content: '';
    display: block;
    background: $white;
    height: 0.5rem;
    width: 0.5rem;
    transform: translateX(-50%) rotate(45deg);
    position: absolute;
    border-radius: 2px;
    border-left: 1px solid $shadow;
    border-top: 1px solid $shadow;
    top: -0.25rem;
    z-index: 2000;

    right: 0.5rem;
  }

  & main {
    padding: 0.5rem;
  }

  & ul {
    padding: 0;
  }

  & a {
    margin: 0;
    padding: 0 0.5rem;
    white-space: nowrap;
  }
}

@media screen and (max-width: $width + 2rem) {
  .open {
    position: inherit;
  }
  .menuWrapper {
    top: 0;
    left: 0;
    margin: 0;
    width: 100%;
    padding: 1rem;
  }
  div.menu:before {
    display: none;
  }
}
