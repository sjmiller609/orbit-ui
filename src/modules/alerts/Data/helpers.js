import { formErrors } from 'instruments'

const errors = [
  {
    key: 'permissions',
    name: 'properties.alert_emails',
    error: 'Contact your workspace admin to upgrade your permissions.',
  },
]

export const handleError = error => formErrors({ error, errors })

export const trimError = error => error.split(':')[1].trim()
