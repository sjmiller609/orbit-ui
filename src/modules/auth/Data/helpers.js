import React from 'react'

import { formErrors, Link } from 'instruments'

const errors = [
  {
    key: 'invalid password',
    name: 'password',
    error: 'Incorrect password',
  },
  {
    key: 'not found',
    name: 'email',
    error: "Hmm, we can't find that email",
  },
  {
    key: 'email already in use',
    name: 'email',
    error: 'That email is already taken.',
  },
  {
    key: 'no password credentials found',
    name: 'email',
    error: 'No password found. Did you mean to login with OAuth?',
  },
  {
    key: 'awaiting email confirmation',
    name: 'email',
    error: (
      <Link to="/resend">
        Your email is not yet verified. Please check your email for a
        verification link.<br /> [Click to resend]
      </Link>
    ),
  },
  {
    key: 'not associated with the specified invite',
    name: 'email',
    error: 'The email you entered does not match your invitation.',
  },
  {
    key: 'permissions',
    name: 'permissions',
    error: 'Contact your Workspace Admin to upgrade your permissions.',
  },
  {
    key: 'Email address in use',
    name: 'email',
    error: (
      <Link to="/login">
        A user with that email address already exists. Click here to login.
      </Link>
    ),
  },
]

export const handleError = error => formErrors({ error, errors })
