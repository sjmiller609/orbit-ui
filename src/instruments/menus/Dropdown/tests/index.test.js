import React from 'react'
import { mountWrap } from 'helpers/tests'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import Dropdown from '../index'

jest.useFakeTimers()

const props = {
  right: false,
  disable: false,
  selector: <p>Select</p>,
  className: 'testClassName',
  small: false,
}

describe('Dropdown', () => {
  it('should be defined', () => {
    expect(Dropdown).toBeDefined()
  })

  it('should render correctly', () => {
    const wrapper = shallow(
      <Dropdown {...props}>
        <p>Children</p>
      </Dropdown>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should render right', () => {
    const wrapper = mountWrap(
      <Dropdown {...props} right={true}>
        <p>Children</p>
      </Dropdown>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should render disabled', () => {
    const wrapper = mountWrap(
      <Dropdown {...props} disable={true}>
        <p>Children</p>
      </Dropdown>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should render small', () => {
    const wrapper = mountWrap(
      <Dropdown {...props} small={true}>
        <p>Children</p>
      </Dropdown>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should handle componentWillUpdate', () => {
    const wrapper = renderer.create(
      <Dropdown {...props}>
        <p>Children</p>
      </Dropdown>
    )
    const inst = wrapper.getInstance()
    const func = jest.fn(() => inst.componentWillUpdate())
    func()
    expect(func).toHaveBeenCalled()
  })

  it('should handle blur', () => {
    const wrapper = renderer.create(
      <Dropdown {...props}>
        <p>Children</p>
      </Dropdown>
    )
    const inst = wrapper.getInstance()
    const func = jest.fn(() => inst.blur())
    func()
    expect(func).toHaveBeenCalled()
  })

  it('should handle open', () => {
    const wrapper = renderer.create(
      <Dropdown {...props}>
        <p>Children</p>
      </Dropdown>
    )
    const inst = wrapper.getInstance()
    const func = jest.fn(() => inst.open())
    func()
    expect(func).toHaveBeenCalled()
  })

  it('should handle timer', () => {
    Dropdown.prototype.unload = jest.fn(() => {
      clearTimeout(this.timeout)
    })

    let wrapper = shallow(
      <Dropdown {...props}>
        <p>Children</p>
      </Dropdown>
    )

    jest.runTimersToTime(1200)
    wrapper.unmount()
  })
})
