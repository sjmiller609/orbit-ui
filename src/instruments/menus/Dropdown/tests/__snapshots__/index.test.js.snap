// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render correctly 1`] = `
<div
  className="dropdown left out testClassName"
>
  <Backdrop
    close={[Function]}
    show={false}
  />
  <Link
    className="selector"
    onClick={[Function]}
  >
    <p>
      Select
    </p>
  </Link>
  <div
    className="menuWrapper"
    onClick={[Function]}
  >
    <Card
      className="menu"
    >
      <p>
        Children
      </p>
    </Card>
  </div>
</div>
`;

exports[`Dropdown should render disabled 1`] = `
<Dropdown
  className="testClassName"
  disable={true}
  right={false}
  selector={
    <p>
      Select
    </p>
  }
  small={false}
>
  <div
    className="dropdown left out testClassName"
  >
    <Backdrop
      close={[Function]}
      show={false}
    />
    <p>
      Select
    </p>
    <div
      className="menuWrapper"
      onClick={[Function]}
    >
      <Card
        className="menu"
      >
        <Box
          className="card menu"
          justify="flex-start"
        >
          <Flex
            className="card menu"
            flow="box"
            justify="flex-start"
          >
            <div
              className="flex box flex-start card menu"
            >
              <main
                className="full"
              >
                <p>
                  Children
                </p>
              </main>
            </div>
          </Flex>
        </Box>
      </Card>
    </div>
  </div>
</Dropdown>
`;

exports[`Dropdown should render right 1`] = `
<Dropdown
  className="testClassName"
  disable={false}
  right={true}
  selector={
    <p>
      Select
    </p>
  }
  small={false}
>
  <div
    className="dropdown right out testClassName"
  >
    <Backdrop
      close={[Function]}
      show={false}
    />
    <Link
      className="selector"
      onClick={[Function]}
    >
      <A
        className="link selector"
        onClick={[Function]}
      >
        <a
          className="link selector"
          onClick={[Function]}
        >
          <p>
            Select
          </p>
        </a>
      </A>
    </Link>
    <div
      className="menuWrapper"
      onClick={[Function]}
    >
      <Card
        className="menu"
      >
        <Box
          className="card menu"
          justify="flex-start"
        >
          <Flex
            className="card menu"
            flow="box"
            justify="flex-start"
          >
            <div
              className="flex box flex-start card menu"
            >
              <main
                className="full"
              >
                <p>
                  Children
                </p>
              </main>
            </div>
          </Flex>
        </Box>
      </Card>
    </div>
  </div>
</Dropdown>
`;

exports[`Dropdown should render small 1`] = `
<Dropdown
  className="testClassName"
  disable={false}
  right={false}
  selector={
    <p>
      Select
    </p>
  }
  small={true}
>
  <div
    className="dropdown left out small testClassName"
  >
    <Backdrop
      close={[Function]}
      show={false}
    />
    <Link
      className="selector"
      onClick={[Function]}
    >
      <A
        className="link selector"
        onClick={[Function]}
      >
        <a
          className="link selector"
          onClick={[Function]}
        >
          <p>
            Select
          </p>
        </a>
      </A>
    </Link>
    <div
      className="menuWrapper"
      onClick={[Function]}
    >
      <Card
        className="menu"
      >
        <Box
          className="card menu"
          justify="flex-start"
        >
          <Flex
            className="card menu"
            flow="box"
            justify="flex-start"
          >
            <div
              className="flex box flex-start card menu"
            >
              <main
                className="full"
              >
                <p>
                  Children
                </p>
              </main>
            </div>
          </Flex>
        </Box>
      </Card>
    </div>
  </div>
</Dropdown>
`;
